### ISSUE TITLE
Save redline PDF for Ministry Sign Off

### ISSUE THREAD

**Issue by lmullane** [bcgov/foi-flow#3437]
* As an IAO analyst
* I want to be able to save a redline PDF with all redactions
* So that I can save the ministry coordinator can download the redline for Ministry Sign Off

**Assumptions & Scope**	
What are the assumptions for this story?

- For release 9.0, the ministry will not have access to the Records Review app.
-  The request must be in the Records Review, Ministry Sign Off or Peer Review state.
- Every page in the request package must have at least one of these flags applied: Partial Disclosure, Full Disclosure, Withheld in Full, Duplicate, Not Responsive, Duplicate or Non Responsive 
- Packages with pages with no flag or the 'Consult', 'In Progress' or 'Page Left Off' flag cannot be saved as a redline
- All redactions will be red boxes with the content visible.
- All records will be saved in one PDF document in Object Storage.
- And all Duplicate and Not Responsive flagged pages will be removed from the redline PDF
- The PDF document will include all redactions, applied  with a red box around the visible content, with the section stamp.
- The PDF document will include NO comments, highlights or annotations. 
- The PDF document will be viewable in the redaction app.

What is IN scope?
Saving the redline as PDF in Object Storage.

Create Records Package dropdown

What is NOT in scope?
Creating redaction layers and toggling between redaction layers is covered in #3704; Creating the Redline layer is done by default when Records are uploaded into the Records Log.
 
State changes.

Saved phased releases of the redline will be in future story.

Creating a new records package.

**Acceptance Criteria**	

**Scenario 1: Active "Create Records Package" Dropdown**
* GIVEN an analyst is on a request
* WHEN IAO analyst is in the redaction app viewing the records
* THEN there will be an active "Create Records Package" dropdown
https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/981498460

**Scenario 2: Click on "Create Records Package" Dropdown**
* GIVEN an IAO analyst is in the redaction app viewing the records
* WHEN IAO analyst clicks on the "Create Records Package" Dropdown
* THEN the dropdown will open
* AND there will 2 options: 'Redline for Sign Off,' 'Response Package for Applicant.'

**Scenario 3: Inactive "Redline for Sign Off"**
* GIVEN a records package is not in Records Review, Peer Review or Ministry Sign Off
* AND every page in the Records Package has one of these page flags applied: Partial Disclosure, Full Disclosure, Withheld in Full, Duplicate, Not Responsive
* WHEN the IAO analyst is clicks on the "Create Records Package" Dropdown
https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/981498459
* THEN there will be inactive and greyed out "Redline for Sign Off" option in the dropdown
https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/981498460

**Scenario 4: Inactive 'Redline for Sign Off"**
* GIVEN a records package is in Records Review, Ministry Sign Off or Peer Review
* AND every page in the Records Package does not have one of these page flags applied: Partial Disclosure, Full Disclosure, Withheld in Full, Duplicate, Not Responsive
* WHEN the IAO analyst is clicks on the "Create Records Package" Dropdown
https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/981498459
* THEN there will be inactive and greyed out "Redline for Sign Off" option in the dropdown
https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/981498460

**Scenario 5: Active 'Redline for Sign Off'**
* GIVEN a records package is in Records Review, Ministry Sign Off or Peer Review
* AND every page in the Records Package has one of these page flags applied: Partial Disclosure, Full Disclosure, Withheld in Full, Duplicate, Not Responsive
* WHEN the IAO analyst clicks on the "Create Records Package" dropdown
* THEN the "Redline for Sign Off" will be clickable
* AND the "Response Package for Applicant" will not be clickable

**Scenario 6: Click "Redline for Sign Off"**
* GIVEN the "Create Records Package" dropdown is open
* AND the request is in Records Review, Ministry Sign Off or Peer Review
https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/981498459
* WHEN the IAO analyst clicks the active "Redline for Sign Off"
* THEN a modal will open with the message, "Are you sure want to create the redline PDF for ministry sign off?". 
* AND a "Create Redline PDF" and "Cancel" button
https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/981498459

**Scenario 7: Click "Cancel" button**
* GIVEN an analyst is on the "Redline for Sign Off" modal
* WHEN the IAO analyst clicks the "Cancel" button
* THEN a modal will close
* AND the analyst will remain in the redaction app
* AND no redline PDF will be saved

**Scenario 8: Click "Create Redline PDF" button**
* GIVEN an analyst is on the "Redline for Sign Off" modal
* WHEN the IAO analyst clicks the "Create Redline PDF" button
* THEN a modal will close
* AND a green success toast will appear when the save is complete
* AND the analyst will remain in the redaction app
* AND the analyst will see the redline layer in the redaction app so they can review if needed
* AND one redline PDF with all records will be saved in Object Storage
* AND all pages flagged as Duplicate or Non Responsive pages will be removed from the redline
* AND the redline PDF will contain all redactions (red boxes around visible text with the section of the ACT stamped in the box)
* AND the redline PDF will contain no comments or annotations of any kind

**Dependencies? What is the impact of this dependency? (If so, link dependency in the ticket, make it visible in a team´s backlog)**

**Validation Rules? (If yes, list here)**	

**Design**
@xxx - please link the Design here

**Definition of Ready**

1. [ ] Is there a well articulated User Story?
2. [ ] Is there Acceptance Criteria that covers all scenarios (happy/sad paths)?
3. [ ] If there is a user interface, is there a design? 
4. [ ] Does the user story need user research/validation?
5. [ ] Does this User Story needs stakeholder approval?
6. [ ] Design / Solution accepted by Product Owner
7. [ ] Is this user story small enough to be completed in a Sprint? Should it be split?
8. [ ] Are the dependencies known/ understood? (technical, business, regulatory/policy)
9. [ ] Has the story been estimated?

**Definition of Done**

1.    [ ] Passes developer unit tests
2.    [ ] Passes peer code review
3.    [ ] If there's a user interface, passes UX assurance
4.    [ ] Passes QA of Acceptance Criteria with verification in Dev and Test
5.	[ ] Confirm Test cases built and succeeding 
6.	[ ] No regression test failures
7.	[ ] Test coverage acceptable by Product Owner
8.	[ ] Ticket ready to be merged to master or story branch
9.	[ ] Developer to list Config changes/ Update documents and designs
10.	[ ] Can be demoed in Sprint Review 
11.	[ ] Tagged as part of a Release
12.   [ ] Feature flagged if required
13.  [ ] Change Management activities done?

**Comment by lmullane** [comment]
@m-prodan, @liseandtea, do redlines include pages that were flagged as Duplicate or Not Responsive?

**Comment by lmullane** [comment]
@JHarrietha-AOT, @m-prodan, @liseandtea, @KyEggleston   

Please review and comment on the AC. Follows the same pattern as saving the records package (#3231).

cc @abin-aot

**Comment by lmullane** [comment]
@m-prodan, as confirmed that pages flagged as Duplicate and Not Responsive will be removed from the redline.

**Comment by JHarrietha-AOT** [comment]
@lmullane @m-prodan 

I have combined the designs for #3231 & #3437. I am hoping to avoid adding multiple buttons thats don't have a CTA that is used throughout the redaction process. It is something that is used to generate the files at specific times. I noticed this in the 4 tickets that have been these quick turn around designs - they were all to add additional buttons. When adding a button for each action isn't always the best solution, so to help ensure that we leave room for growth on the screen and also don't confuse users with way to main primary and secondary buttons like AXIS has. 

Let me know if you have any questions or comments:

ACTIVE: https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/981498459

INACTIVE: https://jacklyn808742.invisionapp.com/console/share/8QVPHCNXSGA/981498460 
Because user has note assigned flags to every page

**Comment by lmullane** [comment]
Thanks for the updated designs, @JHarrietha-AOT. Dropdown is great.

I updated the AC accordingly.

**Comment by lmullane** [comment]
As per demo today, we need to update the AC for this story to include a document status report with a list of pages that have different page flags (e.g. Withheld in Full, Full Disclosure, etc).

**Comment by lmullane** [comment]
Updated the assumptions that a redline cannot be saved if any pages have an 'In Progress' flag.

**Comment by lmullane** [comment]
@m-prodan, please review for clarity.

I updated the AC to include an option to create a Consult PDF in the Create Record Package dropdown.

For simplicity, the dropdown will now always be active, but the options will be greyed out depending on other actions taken by the user.

A new story will be written for Saving a Consult.

@JHarrietha-AOT, we will need an updated design.

cc @divyav-aot,

**Comment by lmullane** [comment]
@JHarrietha-AOT, I updated the AC to say "Response Package for Applicant" instead of "Final Package for Applicant" in the dropdown to be consistent with #3704.

Please update the designs when you have time.

**Comment by JHarrietha-AOT** [comment]
Thanks for flagging this, I have updated the design here:https://jacklyn808742.invisionapp.com/overview/Redaction-cla8feezv003o015h7op82c5c/screens let me know if there is anything else

**Comment by lmullane** [comment]
Note, the Redline will be the default layer in the Records Review app. 

This story is to save the Redline layer as a PDF.

**Comment by richard-aot** [comment]
@lmullane  When saving redline for ministry signoff, do we want to save a merged pdf file for each of the divisions?  And then we can have a backend process creating one zip file for each division, including this merged pdf file and any incompatible files for that division. Is it possible to create a new ticket for this pdf stitching (merge) and zip task? Thanks.

**Comment by KyEggleston** [comment]
Just regarding Scenario 1 and 2 - currently, the dropdown looks as the following:

![image](https://github.com/bcgov/foi-flow/assets/117312323/15c8b905-2270-475f-8dc7-03ae2a0ac28b)

I believe with all the changes we have made to this ticket/story, the dropdown should appear as the following:

![image](https://github.com/bcgov/foi-flow/assets/117312323/968d7735-6f11-46ab-8ee6-3dd2ddb780bc)

**Comment by KyEggleston** [comment]
Updated ACs to remove the reference to being able to create a 'Consult Public Bodies' package, as this is no longer in scope for this ticket.

**Comment by KyEggleston** [comment]
@richard-aot - made some comments/updates after testing this, see above. May want to confirm the second comment around the 'peer Review' state with Matt and Lise at standup. Will wait to QA Scenario 8 when #3301 is done.

**Comment by richard-aot** [comment]
Thanks @KyEggleston. I'm working on the updates and fixes right now. It'll be ready for testing when I push #3231 to test-marshal.

**Comment by KyEggleston** [comment]
From standup - the redline should also be able to be created/saved from the 'Peer Review' state. Will update OP

**Comment by richard-aot** [comment]
The above bugs are fixed and updates are applied.

Scenario 8 is still blocked by #3301.

**Comment by nimya-aot** [comment]
currently if a page inside the package has "Page Left off" flag then also we are not able  to save as redline. Is it missing on the assumptions listed above? or is it a bug? cc @richard-aot 
@lmullane @KyEggleston 
![image.png](https://images.zenhubusercontent.com/63e6c81be9adb8e6f876c7e0/b1594c01-9107-4ec5-8ee0-32b1b341a77a)

**Comment by nimya-aot** [comment]
I noticed that notes are appearing on the redline pdf generated. @richard-aot Could you please check that? This is on request EDU-2023-1108 on test-marshal
 ![image.png](https://images.zenhubusercontent.com/63e6c81be9adb8e6f876c7e0/23cbdb3c-68f8-4e7d-81d4-c4484ec047b4)

**Comment by nimya-aot** [comment]
@richard-aot Needs to fix this issue as well https://app.zenhub.com/workspaces/foi-flow-607f0ab6ae54b700111dc733/issues/gh/bcgov/foi-flow/4265

**Comment by richard-aot** [comment]
@nimya-aot, the 'redact' note in redline pdf issue seems fixed. I tried it on a new/clean request, 'redact' notes are not showing any more. Could you please verify? Thanks.
And for 4265, it is not related to saving redline. The issue is caused by a JPG file. If you upload any other files after saving redline, the issue is gone.

**Comment by nimya-aot** [comment]
@richard-aot It still showing ' Redact' note on one of the redactions. Could you please check?
 [Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBcEE1IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--fe2902bb8bc948ea4be1615f909c8f93de037d77/image.png)

Also on another place it shows 'Textbox' note. 

[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBcFE1IiwiZXhwIjpudWxsLCJwdXIiOiJibG9iX2lkIn19--2eb1041fc792522f7b8f709478d546f7d4324361/image.png)

This is for request #EDU-2023-1014. Also if we change any file, its not updating on the redline package stored on s3 browser.

**Comment by KyEggleston** [comment]
@richard-aot - one assumption / AC that was missing from this ticket. 

When a 'Redline for Sign Off' is saved, and then downloadable through the 'Records' tab - if a program area/division has only provided incompatible files (i.e., they do not have a redline PDF, and none of their files were uploaded into the redaction app), we should still produce a package for sign off containing those incompatible files for the program area/division. 

For example, a division only provides incompatible files. These files would show up as individual files for the division when the Harms package is created/downloaded. These files would not show up in the Redaction App, and therefore, would not be redlined/redacted in the Redaction App. 

When the user selects to save the 'Redline for Sign Off' in the Redaction App, if a division has only provided incompatible files, these should also be included in the 'Redline for Sign Off' package, contained within a folder for the division. This would look similar to the Harms package. 

![image](https://github.com/bcgov/foi-flow/assets/117312323/b0a5f670-cdd8-4cc2-9b7b-c4cee24e4312)

![image](https://github.com/bcgov/foi-flow/assets/117312323/8bd2d26e-a7c9-47d4-bf3b-b7474be4aa48)

**Comment by divyav-aot** [comment]
@KyEggleston I can take a look at this issue.

**Comment by divyav-aot** [comment]
@KyEggleston - the issue mentioned related to incompatible files are fixed and ready for QA.

**Comment by KyEggleston** [comment]
Scenario 8 is failing - pages flagged with 'Not Responsive' and 'Duplicate' ARE being included in the Redline for Ministry Sign Off. 

Pages flagged with 'Not Responsive' and 'Duplicate' should be removed from the Redline for Ministry Sign Off.

**Comment by liseandtea** [comment]
All AC pass - can close ticket. Double checked on scenario 8, Not Responsive' and 'Duplicate' are not being included in the Redline for Ministry Sign Off.
